src/autoscaler/metricsforwarder/server/custom_metrics_handlers.go

```diff
@@ -0,0 +1,208 @@
+		policyDB:           policyDB,
+		credentialCache:    credentialCache,
+		allowedMetricCache: allowedMetricCache,
+		cacheTTL:           cacheTTL,
+		logger:             logger,
+	}
+}
+
+func (mh *CustomMetricsHandler) PublishMetrics(w http.ResponseWriter, r *http.Request, vars map[string]string) {
+	w.Header().Set("Content-Type", "application/json")
+
+	username, password, authOK := r.BasicAuth()
+
+	if authOK == false {
+		mh.logger.Info("error-processing-authorizaion-header")
+		handlers.WriteJSONResponse(w, http.StatusUnauthorized, models.ErrorResponse{
+			Code:    "Authorization-Failure-Error",
+			Message: "Error verifying user credentials. Basic authorization is not used properly"})
+		return
+	}
+
+	var isValid bool
+	var credentials models.CustomMetricCredentials
+
+	appID := vars["appid"]
+	res, found := mh.credentialCache.Get(appID)
+	if found {
+		// Credentials found in cache
+		credentials = res.(models.CustomMetricCredentials)
+		isValid = mh.validateCredentials(username, credentials.Username, password, credentials.Password)
+	}
+
+	// Credentials not found in cache or
+	// stale cache entry with invalid credential found in cache
+	// search in the database and update the cache
+	if !found || !isValid {
+		usernameHash, passwordHash, err := mh.policyDB.GetCustomMetricsCreds(appID)
+		if err != nil {
+			mh.logger.Error("error-during-getting-binding-credentials-from-policyDB", err)
+			handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+				Code:    "Interal-Server-Error",
+				Message: "Error reading custom metrics request body"})
+			return
+		}
+		credentials = models.CustomMetricCredentials{
+			Username: usernameHash,
+			Password: passwordHash,
+		}
+		// update the cache
+		mh.credentialCache.Set(appID, credentials, mh.cacheTTL)
+
+		isValid = mh.validateCredentials(username, credentials.Username, password, credentials.Password)
+		// If Credentials in DB is not valid
+		if !isValid {
+			mh.logger.Error("error-validating-authorizaion-header", err)
+			handlers.WriteJSONResponse(w, http.StatusUnauthorized, models.ErrorResponse{
+				Code:    "Authorization-Failure-Error",
+				Message: "Basic authorization credential does not match"})
+			return
+		}
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		mh.logger.Error("error-reading-request-body", err, lager.Data{"body": r.Body})
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error reading custom metrics request body"})
+		return
+	}
+	var metricsConsumer *models.MetricsConsumer
+	err = json.Unmarshal(body, &metricsConsumer)
+	if err != nil {
+		mh.logger.Error("error-unmarshaling-metrics", err, lager.Data{"body": r.Body})
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error unmarshaling custom metrics request body"})
```

Comment:
If the submitted custom metric doesn't align with the expected format of MetricsConsumer,  the json.unmarshal here will fail.  In this case,  should we should Interal-Server-Error ?  I think it should be 400 Bad request.

Reply:
Making sense, changed to BadRequest